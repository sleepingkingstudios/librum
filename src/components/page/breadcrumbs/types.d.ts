export type Breadcrumb = {
  active?: true | false;
  label: string;
  url?: string;
};
