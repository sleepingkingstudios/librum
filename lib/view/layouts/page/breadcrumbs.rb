# frozen_string_literal: true

module View::Layouts
  # Component for rendering the page breadcrumb navigation.
  class Page::Breadcrumbs < ViewComponent::Base
    # Configuration object for a page breadcrumb.
    class BreadcrumbDefinition
      # @param label [String] the label to display for the breadcrumb.
      # @param url [String] the breadcrumb url.
      # @param active [true, false] if true, marks the breadcrumb as active and
      #   disables the link. Defaults to false.
      def initialize(label:, url:, active: false)
        @active = active
        @label  = label
        @url    = url
      end

      # @return [true, false] if true, marks the breadcrumb as active and
      #   disables the link.
      attr_reader :active

      # @return [String] the label to display for the breadcrumb.
      attr_reader :label

      # @return [String] the breadcrumb url.
      attr_reader :url
    end

    # @param breadcrumbs
    #   [Array<View::Layouts::Page::Breadcrumbs::BreadcrumbConfiguration>] the
    #   breadcrumbs to render.
    def initialize(breadcrumbs:)
      super()

      @breadcrumbs = (breadcrumbs || []).map do |breadcrumb|
        next breadcrumb if breadcrumb.is_a?(BreadcrumbDefinition)

        BreadcrumbDefinition.new(**breadcrumb)
      end
    end

    # @return [Array<View::Layouts::Page::Breadcrumbs::BreadcrumbConfiguration>]
    #   the breadcrumbs to render.
    attr_reader :breadcrumbs
  end
end
