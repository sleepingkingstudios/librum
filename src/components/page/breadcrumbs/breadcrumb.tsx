import * as React from 'react';
import {
  Link,
  useLocation,
} from 'react-router-dom';

import type { Breadcrumb } from './types';

const EmptyBreadcrumb = ({ label }: { label: string }): JSX.Element => (
  <li className="page-breadcrumb">
    <span className="page-breadcrumb-empty">{ label }</span>
  </li>
);

const BreadcrumbLink = (
  { label, url }: { label: string, url: string }
): JSX.Element => (
  <li className="page-breadcrumb">
    <Link to={url} className="page-breadcrumb-link">{ label }</Link>
  </li>
)

export const PageBreadcrumb = (
  { breadcrumb }: { breadcrumb: Breadcrumb }
): JSX.Element => {
  const { active, label } = breadcrumb;
  const { pathname } = useLocation();

  if (!('url' in breadcrumb) || active === false) {
    return (<EmptyBreadcrumb label={label} />);
  }

  const { url } = breadcrumb;

  if (active !== true && pathname === url) {
    return (<EmptyBreadcrumb label={label} />);
  }

  return (<BreadcrumbLink label={label} url={url} />);
};
