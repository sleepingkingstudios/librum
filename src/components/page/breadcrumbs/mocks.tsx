import * as React from 'react';

import type { Breadcrumb } from './types';

const formatBreadcrumb = (breadcrumb: Breadcrumb): string => {
  const { active, label, url } = breadcrumb;

  return `${label}${url ? ` @ ${url}` : ''}${active ? ' (active)' : ''}`;
};

export const PageBreadcrumbs = ({
  breadcrumbs,
}: {
  breadcrumbs: Breadcrumb[],
}): JSX.Element => (
  <ul aria-label="breadcrumbs">
    {
      breadcrumbs.map((breadcrumb: Breadcrumb) => (
        <li aria-label="breadcrumb" key={breadcrumb.label}>
          { formatBreadcrumb(breadcrumb) }
        </li>
      ))
    }
  </ul>
);
