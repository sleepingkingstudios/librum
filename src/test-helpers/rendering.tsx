import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
