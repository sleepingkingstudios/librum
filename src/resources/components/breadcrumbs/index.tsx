import * as React from 'react';

import { kebabCase } from 'lodash';
import { useParams } from 'react-router-dom';

import { applyWildcards } from '@api/utils';
import type { Breadcrumb } from '@components/page';
import { PageBreadcrumbs } from '@components/page/breadcrumbs';
import { titleCase } from '@utils/text';

const generatePageBreadcrumbs = ({
  action,
  member,
  page,
  resourceBreadcrumbs,
  resourceName,
}: {
  action: string,
  member: boolean,
  page: {
    breadcrumbs?: Breadcrumb[],
  },
  resourceBreadcrumbs: Breadcrumb[],
  resourceName: string,
}): Breadcrumb[] => {
  if (page.breadcrumbs) {
    return [...resourceBreadcrumbs, ...page.breadcrumbs];
  }

  const resourceUrl = joinUrl(
    lastUrl(resourceBreadcrumbs),
    kebabCase(resourceName),
  );
  const breadcrumbs = [...resourceBreadcrumbs];

  breadcrumbs.push({
    label: titleCase(resourceName),
    url: resourceUrl,
  });

  if (member) {
    // @TODO
  } else {
    breadcrumbs.push({
      active: true,
      label: titleCase(action),
    });
  }

  return breadcrumbs;
};

const generateResourceBreadcrumbs = ({
  breadcrumbs,
  scope,
}: {
  breadcrumbs?: Breadcrumb[],
  scope?: string,
}): Breadcrumb[] => {
  if (breadcrumbs) { return breadcrumbs; }

  const prefix: Breadcrumb[] = [
    {
      label: 'Home',
      url: '/',
    },
  ];

  if (!(scope && scope.length > 0)) { return prefix; }

  scope.split('/').forEach((segment: string) => {
    const previous: Breadcrumb = prefix[prefix.length - 1];
    const { url } = previous;

    prefix.push({
      label: titleCase(segment),
      url: joinUrl(url, kebabCase(segment)),
    });
  });

  return prefix;
};

const joinUrl = (prefix: string, suffix: string): string => {
  if (prefix[prefix.length - 1] === '/') {
    return `${prefix}${suffix}`;
  }

  return `${prefix}/${suffix}`;
};

const lastUrl = (breadcrumbs: Breadcrumb[]): string => {
  if (breadcrumbs.length === 0) { return '/'; }

  return breadcrumbs.reduce(
    (memo: string, breadcrumb: Breadcrumb) => {
      const { url } = breadcrumb;

      return url ? url : memo;
    },
    '/',
  );
};

export const ResourceBreadcrumbs = ({
  action,
  breadcrumbs,
  member,
  page,
  resourceName,
  scope,
  wildcards,
}: {
  action: string,
  breadcrumbs?: Breadcrumb[],
  member: boolean,
  page: { breadcrumbs?: Breadcrumb[] },
  resourceName: string,
  scope?: string,
  wildcards?: Record<string, string>,
}): JSX.Element => {
  const resourceBreadcrumbs: Breadcrumb[] =
    generateResourceBreadcrumbs({ breadcrumbs, scope });
  const pageBreadcrumbs: Breadcrumb[] =
    generatePageBreadcrumbs({
      action,
      member,
      page,
      resourceBreadcrumbs,
      resourceName,
    });
  const params = useParams();

  const appliedBreadcrumbs: Breadcrumb[] = pageBreadcrumbs.map(
    (breadcrumb: Breadcrumb): Breadcrumb => {
      if (!('url' in breadcrumb)) { return breadcrumb; }

      return {
        ...breadcrumb,
        url: applyWildcards({
          url: breadcrumb.url,
          wildcards: { ...params, ...wildcards },
        }),
      };
    }
  );

  return (<PageBreadcrumbs breadcrumbs={appliedBreadcrumbs} />);
};
