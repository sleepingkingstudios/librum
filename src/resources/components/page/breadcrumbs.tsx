import * as React from 'react';
import { isPlainObject } from 'lodash';

import { kebabCase } from 'lodash';
import { useParams } from 'react-router-dom';

import { applyWildcards } from '@api/utils';
import type { Breadcrumb } from '@components/page';
import { PageBreadcrumbs } from '@components/page/breadcrumbs';
import { singularize } from '@utils/inflector';
import { titleCase } from '@utils/text';

const generateCollectionBreadcrumbs = ({
  action,
  resourceBreadcrumbs,
  resourceName,
}: {
  action: string,
  resourceBreadcrumbs: Breadcrumb[],
  resourceName: string,
}): Breadcrumb[] => {
  if (action === '') {
    return [
      {
        label: titleCase(resourceName),
        active: true,
      },
    ];
  }

  const resourceUrl = joinUrl(
    lastUrl(resourceBreadcrumbs),
    kebabCase(resourceName),
  );

  return [
    {
      label: titleCase(resourceName),
      url: resourceUrl,
    },
    {
      active: true,
      label: titleCase(action),
    },
  ];
};

const generateMemberBreadcrumbs = ({
  action,
  data,
  resourceBreadcrumbs,
  resourceName,
  singularName,
  status,
}: {
  action: string,
  data: Record<string, unknown>,
  resourceBreadcrumbs: Breadcrumb[],
  resourceName: string,
  singularName?: string,
  status: string,
}): Breadcrumb[] => {
  const resourceUrl = joinUrl(
    lastUrl(resourceBreadcrumbs),
    kebabCase(resourceName),
  );
  const breadcrumbs: Breadcrumb[] = [
    {
      label: titleCase(resourceName),
      url: resourceUrl,
    },
  ];

  if (action === '') {
    if (status !== 'success') {
      return [
        ...breadcrumbs,
        {
          label: 'Loading',
          active: true,
        },
      ];
    }

    return [
      ...breadcrumbs,
      memberBreadcrumbFor({
        active: true,
        data,
        resourceName,
        resourceUrl,
        singularName,
      }),
    ];
  }

  if (status !== 'success') {
    return [
      ...breadcrumbs,
      {
        label: 'Loading',
      },
      {
        label: titleCase(action),
        active: true,
      },
    ];
  }

  return [
    ...breadcrumbs,
    memberBreadcrumbFor({
      active: false,
      data,
      resourceName,
      resourceUrl,
      singularName,
    }),
    {
      label: titleCase(action),
      active: true,
    }
  ];
};

const generatePageBreadcrumbs = ({
  action,
  data,
  page,
  resourceBreadcrumbs,
  resourceName,
  singularName,
  status,
}: {
  action: string,
  data: Record<string, unknown>,
  page: {
    breadcrumbs?: Breadcrumb[],
    member?: boolean,
  },
  resourceBreadcrumbs: Breadcrumb[],
  resourceName: string,
  singularName?: string,
  status: string,
}): Breadcrumb[] => {
  if (page.breadcrumbs) {
    return [...resourceBreadcrumbs, ...page.breadcrumbs];
  }

  let breadcrumbs: Breadcrumb[];

  if (page.member) {
    breadcrumbs = generateMemberBreadcrumbs({
      action,
      data,
      resourceBreadcrumbs,
      resourceName,
      singularName,
      status,
    });
  } else {
    breadcrumbs = generateCollectionBreadcrumbs({
      action,
      resourceBreadcrumbs,
      resourceName,
    });
  }

  return [...resourceBreadcrumbs, ...breadcrumbs];
};

const generateResourceBreadcrumbs = ({
  breadcrumbs,
  scope,
}: {
  breadcrumbs?: Breadcrumb[],
  scope?: string,
}): Breadcrumb[] => {
  if (breadcrumbs) { return breadcrumbs; }

  const prefix: Breadcrumb[] = [
    {
      label: 'Home',
      url: '/',
    },
  ];

  if (!(scope && scope.length > 0)) { return prefix; }

  scope.split('/').forEach((segment: string) => {
    const previous: Breadcrumb = prefix[prefix.length - 1];
    const { url } = previous;

    prefix.push({
      label: titleCase(segment),
      url: joinUrl(url, kebabCase(segment)),
    });
  });

  return prefix;
};

const joinUrl = (prefix: string, suffix: string): string => {
  if (prefix[prefix.length - 1] === '/') {
    return `${prefix}${suffix}`;
  }

  return `${prefix}/${suffix}`;
};

const lastUrl = (breadcrumbs: Breadcrumb[]): string => {
  if (breadcrumbs.length === 0) { return '/'; }

  return breadcrumbs.reduce(
    (memo: string, breadcrumb: Breadcrumb) => {
      const { url } = breadcrumb;

      return url ? url : memo;
    },
    '/',
  );
};

const memberBreadcrumbFor = ({
  active,
  data,
  resourceName,
  resourceUrl,
  singularName,
}: {
  active: boolean,
  data: Record<string, unknown>,
  resourceName: string,
  resourceUrl: string,
  singularName?: string,
}): Breadcrumb => {
  const resourceKey = singularName || singularize(resourceName);

  if (resourceKey in data && isPlainObject(data[resourceKey])) {
    const resource = data[resourceKey] as Record<string, unknown>;

    if ('name' in resource) {
      const { name } = resource;
      const key = resourceKeyFor(resource);

      if (active) { return { active, label: name.toString() }; }

      if (key === null || key.length === 0) {
        return { label: name.toString() };
      }

      return {
        label: name.toString(),
        url: `${resourceUrl}/${key}`,
      };
    }
  }

  if (active) { return { active, label: titleCase(resourceKey) }; }

  return { label: titleCase(resourceKey) };
};

const resourceKeyFor = (resource: Record<string, unknown>): string => {
  if ('slug' in resource) { return resource.slug.toString(); }

  if ('id' in resource) { return resource.id.toString(); }

  return null;
};

export const ResourcePageBreadcrumbs = ({
  action,
  breadcrumbs,
  data = {},
  page,
  resourceName,
  scope,
  singularName,
  status = 'uninitialized',
  wildcards,
}: {
  action: string,
  breadcrumbs?: Breadcrumb[],
  data?: Record<string, unknown>,
  page: {
    breadcrumbs?: Breadcrumb[],
    member?: boolean,
  },
  resourceName: string,
  scope?: string,
  singularName?: string,
  status?: string,
  wildcards?: Record<string, string>,
}): JSX.Element => {
  const resourceBreadcrumbs: Breadcrumb[] =
    generateResourceBreadcrumbs({ breadcrumbs, scope });
  const pageBreadcrumbs: Breadcrumb[] =
    generatePageBreadcrumbs({
      action,
      data,
      page,
      resourceBreadcrumbs,
      resourceName,
      singularName,
      status,
    });
  const params = useParams();

  const appliedBreadcrumbs: Breadcrumb[] = pageBreadcrumbs.map(
    (breadcrumb: Breadcrumb): Breadcrumb => {
      if (!('url' in breadcrumb)) { return breadcrumb; }

      return {
        ...breadcrumb,
        url: applyWildcards({
          url: breadcrumb.url,
          wildcards: { ...params, ...wildcards },
        }),
      };
    }
  );

  return (<PageBreadcrumbs breadcrumbs={appliedBreadcrumbs} />);
};
