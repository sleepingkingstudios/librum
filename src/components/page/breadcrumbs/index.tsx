import * as React from 'react';

import { PageBreadcrumb } from './breadcrumb';
import type { Breadcrumb } from './types';

export type { Breadcrumb } from './types';

export const PageBreadcrumbs = (
  { breadcrumbs }: { breadcrumbs: Breadcrumb[] }
): JSX.Element => {
  if (breadcrumbs.length === 0) { return null; }

  return (
    <ul className="flex flex-col sm:flex-row page-breadcrumbs">
      {
        breadcrumbs.map((breadcrumb: Breadcrumb): JSX.Element => {
          const { label } = breadcrumb;

          return (
            <PageBreadcrumb key={label} breadcrumb={breadcrumb} />
          );
        })
      }
    </ul>
  )
};
