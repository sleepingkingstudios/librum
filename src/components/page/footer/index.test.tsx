import * as React from 'react';

import '@testing-library/jest-dom';

import { PageFooter } from './index';
import { PageBreadcrumbs } from '../breadcrumbs';
import type { Breadcrumb } from '../breadcrumbs';
import { render } from '@test-helpers/rendering';

describe('<PageFooter>', () => {
  it('should match the snapshot', () => {
    const { asFragment } = render(<PageFooter />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('with breadcrumbs: component', () => {
    const breadcrumbs: Breadcrumb[] = [
      {
        label: 'Home',
        url: '/',
      },
      {
        label: 'Admin'
      },
      {
        label: 'Example',
        url: '/example',
      },
    ];
    const Breadcrumbs =
      (): JSX.Element => (<PageBreadcrumbs breadcrumbs={breadcrumbs} />);

    it('should match the snapshot', () => {
      const { asFragment, getByText } = render(
        <PageFooter breadcrumbs={<Breadcrumbs />} />,
        { router: true }
      );

      breadcrumbs.forEach((breadcrumb: Breadcrumb) => {
        const { label } = breadcrumb;

        expect(getByText(label)).toBeVisible();
      });

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with breadcrumbs: value', () => {
    const breadcrumbs: Breadcrumb[] = [
      {
        label: 'Home',
        url: '/',
      },
      {
        label: 'Admin'
      },
      {
        label: 'Example',
        url: '/example',
      },
    ];

    it('should match the snapshot', () => {
      const { asFragment, getByText } = render(
        <PageFooter breadcrumbs={breadcrumbs} />,
        { router: true }
      );

      breadcrumbs.forEach((breadcrumb: Breadcrumb) => {
        const { label } = breadcrumb;

        expect(getByText(label)).toBeVisible();
      });

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
