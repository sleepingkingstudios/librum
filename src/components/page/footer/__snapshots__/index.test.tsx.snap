// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageFooter> should match the snapshot 1`] = `
<DocumentFragment>
  <footer
    class="mt-3 shrink-0 text-center sm:text-left"
  >
    <hr
      class="hr-fancy mb-2"
    />
    <span
      class="text-muted"
    >
      What lies beyond the furthest reaches of the sky?
    </span>
  </footer>
</DocumentFragment>
`;

exports[`<PageFooter> with breadcrumbs: component should match the snapshot 1`] = `
<DocumentFragment>
  <footer
    class="mt-3 shrink-0 text-center sm:text-left"
  >
    <ul
      class="flex flex-col sm:flex-row page-breadcrumbs"
    >
      <li
        class="page-breadcrumb"
      >
        <span
          class="page-breadcrumb-empty"
        >
          Home
        </span>
      </li>
      <li
        class="page-breadcrumb"
      >
        <span
          class="page-breadcrumb-empty"
        >
          Admin
        </span>
      </li>
      <li
        class="page-breadcrumb"
      >
        <a
          class="page-breadcrumb-link"
          href="/example"
        >
          Example
        </a>
      </li>
    </ul>
    <hr
      class="hr-fancy mb-2"
    />
    <span
      class="text-muted"
    >
      What lies beyond the furthest reaches of the sky?
    </span>
  </footer>
</DocumentFragment>
`;

exports[`<PageFooter> with breadcrumbs: value should match the snapshot 1`] = `
<DocumentFragment>
  <footer
    class="mt-3 shrink-0 text-center sm:text-left"
  >
    <ul
      class="flex flex-col sm:flex-row page-breadcrumbs"
    >
      <li
        class="page-breadcrumb"
      >
        <span
          class="page-breadcrumb-empty"
        >
          Home
        </span>
      </li>
      <li
        class="page-breadcrumb"
      >
        <span
          class="page-breadcrumb-empty"
        >
          Admin
        </span>
      </li>
      <li
        class="page-breadcrumb"
      >
        <a
          class="page-breadcrumb-link"
          href="/example"
        >
          Example
        </a>
      </li>
    </ul>
    <hr
      class="hr-fancy mb-2"
    />
    <span
      class="text-muted"
    >
      What lies beyond the furthest reaches of the sky?
    </span>
  </footer>
</DocumentFragment>
`;
